File: IntersectionLinkedList.java

/**
 * Time Complexity - O(M + N) where N is the length of list headA, M is the length of list headB.
 * Space Complexity - O(1).
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {    
        ListNode ptr1 = headA, ptr2 = headB;
        int len1 = 0, len2 = 0;

        while(ptr1 != null) {
            len1++;
            ptr1 = ptr1.next;
        }
        while(ptr2 != null) {
            len2++;
            ptr2 = ptr2.next;
        }
        if(len1 > len2) {
            ptr1 = headA;
            ptr2 = headB;
            len1 = len1 - len2;
        }
        else {
            ptr1 = headB;
            ptr2 = headA;
            len1 = len2 - len1;
        }
        while(len1 > 0) {
            --len1;
            ptr1 = ptr1.next;
        }
        
        while(ptr1 != null && ptr2 != null) {
            if(ptr1 == ptr2)
                return ptr1;
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }

        return null;
    }
}

File: IntersectionLinkedList.py
# Time Complexity : O(n) for each operation.
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

# Approach is to traverse the higher length list to same level of second low length list and return if two pointers are met.

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        temp1=headA
        lA=0
        while(temp1 is not None):
            temp1=temp1.next
            lA+=1
        temp2=headB
        lB=0
        while(temp2 is not None):
            temp2=temp2.next
            lB+=1
        
        p1=headA
        p2=headB
        while(lA<lB):
            p2=p2.next
            lB-=1
        while(lB<lA):
            p1=p1.next
            lA-=1
        
        while(p1!=None):
            if(p1==p2):
                return p1
            p1=p1.next
            p2=p2.next
            
        return p1